lib/js/protocol/MessageFrameAssembler.js

```diff
@@ -30,55 +30,154 @@
+ * @private {array} _accumulator Combines incoming bytes from consecutive frames.
+ */
+class MessageFrameAssembler {
 
     /**
     * @constructor
     * @param {Number} headerSize
     * @param {Function} callback
     */
-    constructor(headerSize=8, callback){
-        if(callback == null){
+    constructor(callback) {
+        if (callback == null) {
             throw "Callback can't be null for MessageFrameAssembler";
         }
-        
-        this._accumulator = null; //TODO byte array output stream
-        this._headersize = headerSize;
-        this._totalSize = 0;
+
         this._callback = callback;
-    }
-    
-    /**
-    * @param {SdlPacket} sdlPacket
-    */
-    handleFrame(sdlPacket){
 
+        this._accumulator = [];
+        this._totalConsecutiveFrames = 0;
+        this._consecutiveFramesHandledCount = 0;
+        this._consecutiveFramesDataLength = 0;
     }
 
     /**
-    * @param {SdlPacket} sdlPacket
+     * Accepts a single frame or multiframe packet. The callback is used when the packet
+     * is completely read. For a single frame this is immediately. For a multi frame packet
+     * @param {SdlPacket} sdlPacket - Incoming sdl packet to be read.
     */
-    handleMultiFrameMessageFrame(sdlPacket){
-
+    handleFrame(sdlPacket) {
+        const frameType = sdlPacket.getFrameType();
+        //If the sdl packet is a single frame nothing needs to be assembled and this can be returned immediately.
+        if (frameType != FrameType.FIRST && frameType != FrameType.CONSECUTIVE) {
+            return this._callback(null, sdlPacket);
+        }
+        return this._handleMultiFrameMessage(sdlPacket);
     }
 
     /**
-    * @param {SdlPacket} sdlPacket
-    */
-    handleFirstDataFrame(sdlPacket){
+     * Handles the first data frame. Sets the expected frame count and data length.
+     * @param {SdlPacket} sdlPacket 
+     * @private
+     */
+    _handleFirstDataFrame(sdlPacket) {
+
+        /** @type {Uint8Array} */
+        let payload = sdlPacket.getPayload();
+
+        //first 4 8-bit integers contain the data length.
+        let dataLength = (payload[0] & 0x0F) << 24; //8-bit x 3
+        dataLength += (payload[1] & 0xFF) << 16; //8 x 2
+        dataLength += (payload[2] & 0xFF) << 8; //8 x 1
+        dataLength += payload[3] & 0xFF; //8 x 0
 
+        let frameCount = (payload[4] & 0xFF) << 24;
+        frameCount += (payload[5] & 0xFF) << 16;
+        frameCount += (payload[6] & 0xFF) << 8;
+        frameCount += payload[7] & 0xFF;
+
+        this._totalConsecutiveFrames = frameCount;
+        this._consecutiveFramesDataLength = dataLength;
     }
-    
+
     /**
-    * @param {SdlPacket} sdlPacket
-    */
-    handleRemainingFrame(sdlPacket){
+     * First frame should be read in before this is called. Once the last
+     * frame is read, callback is used.
+     * @param {SdlPacket} sdlPacket 
+     */
+    _handleConsecutiveFrame(sdlPacket) {
+        this._consecutiveFramesHandledCount++;
```

Comment:
Instead of counting up a different variable, does it make more sense to take the `_totalConsecutiveFrames` and count down? Then just ensure that it equals 0 when we get the end frame? (Just a question here, not a needed change if you don't agree)

Reply:
I feel that keeping track of the number of frames successfully loaded before the last frame is given could be useful if something unexpected comes up. Especially if it is something like 255 frames loaded when 260 were expected and would indicate that the frame counter did not get set properly after its limit was reached.